src/chainstate/stacks/boot/bns.clar

```diff
@@ -0,0 +1,907 @@
+    (is-eq char 0x69) ;; i
+    (is-eq char 0x6f) ;; o
+    (is-eq char 0x75) ;; u
+    (is-eq char 0x79))) ;; y
+
+(define-private (is-special-char (char (buff 1)))
+  (or 
+    (is-eq char 0x2d) ;; -
+    (is-eq char 0x5f))) ;; _
+
+(define-private (is-char-valid (char (buff 1)))
+  (or 
+    (is-lowercase-alpha char)
+    (is-digit char)
+    (is-special-char char)))
+
+(define-private (is-nonalpha (char (buff 1)))
+  (or 
+    (is-digit char)
+    (is-special-char char)))
+
+(define-private (has-vowels-chars (name (buff 16)))
+  (> (len (filter is-vowel name)) u0))
+
+(define-private (has-nonalpha-chars (name (buff 16)))
+  (> (len (filter is-nonalpha name)) u0))
+
+(define-private (has-invalid-chars (name (buff 19)))
+  (< (len (filter is-char-valid name)) (len name)))
+
+(define-private (compute-name-price (name (buff 16))
+                                    (price-function (tuple (buckets (list 16 uint)) (base uint) (coeff uint) (nonalpha-discount uint) (no-vowel-discount uint))))
+  (let (
+    (exponent (get-exp-at-index (get buckets price-function) (min u15 (- (len name) u1))))
+    (no-vowel-discount (if (not (has-vowels-chars name)) (get no-vowel-discount price-function) u1))
+    (nonalpha-discount (if (has-nonalpha-chars name) (get nonalpha-discount price-function) u1)))
+    (*
+      (/
+        (*
+          (get coeff price-function)
+          (pow (get base price-function) exponent))
+        (max nonalpha-discount no-vowel-discount))
+      u10)))
+
+(define-private (is-name-lease-expired? (namespace (buff 19)) (name (buff 16)))
+  (let (
+    (namespace-props (unwrap! 
+      (map-get? namespaces ((namespace namespace))) 
+      (err ERR_NAMESPACE_NOT_FOUND)))
+    (name-props (unwrap! 
+      (map-get? name-properties ((namespace namespace) (name name))) 
+      (err ERR_NAME_NOT_FOUND))))
+    (if (and (is-none (get registered-at name-props)) (is-some (get imported-at name-props)))
+      (ok false) ;; The name was imported and not launched - not subject to expiration, however should they expire if the namespace expire? - if so, todo(ludo)
+      (let (      ;; The name was registered
+        (registered-at (unwrap! 
+          (get registered-at name-props) (err ERR_PANIC)))
+        (lifetime 
+          (get renewal-rule namespace-props)))
+        (if (is-eq lifetime u0)
+          (ok false)
+          (ok (> block-height (+ lifetime registered-at))))))))
+
+;; todo(ludo): should be refactored - based on (is-name-lease-expired?)
+(define-read-only (is-name-in-grace-period? (namespace (buff 19)) (name (buff 16)))
+  (let (
+    (namespace-props (unwrap! 
+      (map-get? namespaces ((namespace namespace))) 
+      (err ERR_NAMESPACE_NOT_FOUND)))
+    (name-props (unwrap! 
+      (map-get? name-properties ((namespace namespace) (name name))) 
+      (err ERR_NAME_NOT_FOUND))))
+    (if (is-none (get registered-at name-props))
+      (ok false) ;; The name was imported - not subject to expiration
```

Comment:
Imported names are _definitely_ subject to expiration!  Their lifetime begins when the namespace is launched, and they expire once the namespace-prescribed number of blocks pass.

Reply:
Thank you for clarifying - addressed with https://github.com/blockstack/stacks-blockchain/pull/2013/commits/a16f8c8d20520a56095f9133093df8ed97ef81f4